File: components/apps/act-docker/act/mcp_utils/signature_manager.py
"""
ACT Signature Manager

Manages .act.sig files (TOML format) for pre-authenticated node access.
"""

import toml
import os
import re
from typing import Dict, List, Optional, Any
from datetime import datetime


class SignatureManager:
    """Manager for ACT signature files"""

    def __init__(self, signature_path: str):
        """
        Initialize signature manager

        Args:
            signature_path: Path to .act.sig file
        """
        self.signature_path = signature_path
        self.signature: Dict[str, Any] = {}
        self._loaded = False

    def load(self) -> Dict[str, Any]:
        """
        Load and parse signature file

        Returns:
            Parsed signature dictionary

        Raises:
            FileNotFoundError: If signature file doesn't exist
            toml.TomlDecodeError: If TOML syntax is invalid
        """
        if not os.path.exists(self.signature_path):
            raise FileNotFoundError(f"Signature file not found: {self.signature_path}")

        with open(self.signature_path, 'r') as f:
            self.signature = toml.load(f)

        self._loaded = True
        return self.signature

    def save(self) -> None:
        """
        Save signature back to file

        Raises:
            RuntimeError: If signature not loaded
        """
        if not self._loaded:
            raise RuntimeError("Signature must be loaded before saving")

        # Update modified timestamp
        if 'signature' in self.signature:
            self.signature['signature']['updated_at'] = datetime.utcnow().isoformat() + 'Z'

        # Ensure directory exists
        os.makedirs(os.path.dirname(self.signature_path), exist_ok=True)

        with open(self.signature_path, 'w') as f:
            toml.dump(self.signature, f)

    def is_authenticated(self, node_type: str) -> bool:
        """
        Check if a node is authenticated

        Args:
            node_type: Node type (e.g., "github")

        Returns:
            True if authenticated, False otherwise
        """
        node_key = f"node:{node_type}"
        return (node_key in self.signature and
                self.signature[node_key].get('authenticated', False))

    def get_node_auth(self, node_type: str, resolve_env: bool = True) -> Dict[str, Any]:
        """
        Get authentication data for a node

        Args:
            node_type: Node type
            resolve_env: Whether to resolve {{.env.VAR}} references

        Returns:
            Authentication dictionary
        """
        auth_key = f"node:{node_type}.auth"
        auth = self.signature.get(auth_key, {})

        if resolve_env:
            auth = self._resolve_env_vars(auth)

        return auth

    def get_node_defaults(self, node_type: str) -> Dict[str, Any]:
        """
        Get default parameters for a node

        Args:
            node_type: Node type

        Returns:
            Default parameters dictionary
        """
        defaults_key = f"node:{node_type}.defaults"
        return self.signature.get(defaults_key, {})

    def get_authenticated_nodes(self) -> List[str]:
        """
        Get list of all authenticated node types

        Returns:
            List of node type strings
        """
        nodes = []
        for key in self.signature.keys():
            if key.startswith('node:') and not '.' in key.split('node:')[1]:
                node_type = key.split('node:')[1]
                if self.is_authenticated(node_type):
                    nodes.append(node_type)
        return nodes

    def add_node(self, node_type: str, auth: Dict[str, Any],
                 defaults: Optional[Dict[str, Any]] = None,
                 operations: Optional[Dict[str, Any]] = None,
                 metadata: Optional[Dict[str, Any]] = None):
        """
        Add or update a node in signature

        Args:
            node_type: Node type
            auth: Authentication data (will be stored with .env references)
            defaults: Default parameters
            operations: Available operations
            metadata: Additional metadata
        """
        # Add node section
        node_key = f"node:{node_type}"
        self.signature[node_key] = {
            'type': node_type,
            'enabled': True,
            'authenticated': True
        }

        if metadata:
            self.signature[node_key].update(metadata)

        # Add auth section (convert to .env references)
        auth_key = f"node:{node_type}.auth"
        self.signature[auth_key] = self._create_env_references(node_type, auth)

        # Add defaults section
        if defaults:
            defaults_key = f"node:{node_type}.defaults"
            self.signature[defaults_key] = defaults

        # Add operations section
        if operations:
            ops_key = f"node:{node_type}.operations"
            self.signature[ops_key] = operations

        # Update metadata
        if 'metadata' not in self.signature:
            self.signature['metadata'] = {}

        self.signature['metadata']['authenticated_nodes'] = len(self.get_authenticated_nodes())
        self.signature['metadata']['last_updated'] = datetime.utcnow().isoformat() + 'Z'

        self._loaded = True

    def remove_node(self, node_type: str):
        """
        Remove a node from signature

        Args:
            node_type: Node type to remove
        """
        # Remove all sections for this node
        keys_to_remove = [k for k in self.signature.keys() if k.startswith(f"node:{node_type}")]

        for key in keys_to_remove:
            del self.signature[key]

        # Update metadata
        if 'metadata' in self.signature:
            self.signature['metadata']['authenticated_nodes'] = len(self.get_authenticated_nodes())
            self.signature['metadata']['last_updated'] = datetime.utcnow().isoformat() + 'Z'

        self._loaded = True

    def update_defaults(self, node_type: str, defaults: Dict[str, Any]):
        """
        Update default parameters for a node

        Args:
            node_type: Node type
            defaults: New default parameters
        """
        if not self.is_authenticated(node_type):
            raise ValueError(f"Node '{node_type}' is not authenticated")

        defaults_key = f"node:{node_type}.defaults"
        self.signature[defaults_key] = defaults

        self._loaded = True

    def _resolve_env_vars(self, obj: Any) -> Any:
        """
        Recursively resolve {{.env.VARIABLE}} references

        Args:
            obj: Object to resolve (dict, list, str, etc.)

        Returns:
            Resolved object
        """
        if isinstance(obj, dict):
            return {k: self._resolve_env_vars(v) for k, v in obj.items()}
        elif isinstance(obj, list):
            return [self._resolve_env_vars(item) for item in obj]
        elif isinstance(obj, str):
            # Match {{.env.VAR_NAME}}
            pattern = r'\{\{\.env\.([A-Z_][A-Z0-9_]*)\}\}'
            matches = re.findall(pattern, obj)

            if matches:
                for var_name in matches:
                    value = os.getenv(var_name, '')
                    obj = obj.replace(f'{{{{.env.{var_name}}}}}', value)

            return obj
        else:
            return obj

    def _create_env_references(self, node_type: str, auth: Dict[str, Any]) -> Dict[str, str]:
        """
        Convert auth data to .env references

        Args:
            node_type: Node type
            auth: Auth data with actual values

        Returns:
            Auth data with {{.env.VAR}} references
        """
        result = {}
        node_prefix = node_type.upper()

        for key, value in auth.items():
            env_var = f"{node_prefix}_{key.upper()}"
            result[key] = f"{{{{.env.{env_var}}}}}"

            # Set environment variable
            os.environ[env_var] = str(value)

        return result

File: signature-system/mcp/lib/signature-manager.js
/**
 * Signature Manager
 * Manages ACT signature files (.act.sig) for the MCP server
 */

import fs from 'fs/promises';
import path from 'path';
import toml from '@iarna/toml';

export class SignatureManager {
  constructor(signaturePath) {
    this.signaturePath = signaturePath;
    this.signature = null;
    this._loaded = false;
  }

  /**
   * Load and parse signature file
   */
  async load() {
    try {
      const content = await fs.readFile(this.signaturePath, 'utf-8');
      this.signature = toml.parse(content);
      this._loaded = true;
      return this.signature;
    } catch (error) {
      if (error.code === 'ENOENT') {
        throw new Error(`Signature file not found: ${this.signaturePath}`);
      }
      throw new Error(`Failed to parse signature file: ${error.message}`);
    }
  }

  /**
   * Save signature back to file
   */
  async save() {
    if (!this._loaded) {
      throw new Error('Signature must be loaded before saving');
    }

    // Update timestamp
    if (this.signature.signature) {
      this.signature.signature.updated_at = new Date().toISOString();
    }

    const content = toml.stringify(this.signature);
    await fs.writeFile(this.signaturePath, content, 'utf-8');
  }

  /**
   * Check if a node is authenticated
   */
  isAuthenticated(nodeType) {
    const nodeKey = `node:${nodeType}`;
    return this.signature[nodeKey]?.authenticated === true;
  }

  /**
   * Get authentication data for a node
   */
  getNodeAuth(nodeType, resolveEnv = true) {
    const authKey = `node:${nodeType}.auth`;
    let auth = this.signature[authKey] || {};

    if (resolveEnv) {
      auth = this._resolveEnvVars(auth);
    }

    return auth;
  }

  /**
   * Get default parameters for a node
   */
  getNodeDefaults(nodeType) {
    const defaultsKey = `node:${nodeType}.defaults`;
    return this.signature[defaultsKey] || {};
  }

  /**
   * Get available operations for a node
   */
  getOperations(nodeType) {
    const opsKey = `node:${nodeType}.operations`;
    return this.signature[opsKey] || {};
  }

  /**
   * Get metadata for a node
   */
  getNodeMetadata(nodeType) {
    const metaKey = `node:${nodeType}.metadata`;
    return this.signature[metaKey] || {};
  }

  /**
   * Add or update a node in the signature
   */
  async addNode(nodeType, auth, defaults = {}, operations = {}, metadata = {}) {
    // Base node section
    const nodeKey = `node:${nodeType}`;
    this.signature[nodeKey] = {
      type: nodeType,
      enabled: true,
      authenticated: true,
      auth_configured_at: new Date().toISOString()
    };

    // Auth section (store as env references)
    const authKey = `node:${nodeType}.auth`;
    this.signature[authKey] = this._toEnvReferences(nodeType, auth);

    // Defaults section
    if (Object.keys(defaults).length > 0) {
      const defaultsKey = `node:${nodeType}.defaults`;
      this.signature[defaultsKey] = defaults;
    }

    // Operations section
    if (Object.keys(operations).length > 0) {
      const opsKey = `node:${nodeType}.operations`;
      this.signature[opsKey] = operations;
    }

    // Metadata section
    if (Object.keys(metadata).length > 0) {
      const metaKey = `node:${nodeType}.metadata`;
      this.signature[metaKey] = metadata;
    }

    // Update global metadata
    if (!this.signature.metadata) {
      this.signature.metadata = {
        authenticated_nodes: 0,
        unauthenticated_nodes: 0
      };
    }

    // Check if this is a new node
    const existingNodes = this.getAuthenticatedNodes();
    if (!existingNodes.includes(nodeType)) {
      this.signature.metadata.authenticated_nodes =
        (this.signature.metadata.authenticated_nodes || 0) + 1;
    }

    await this.save();
  }

  /**
   * Remove a node from the signature
   */
  async removeNode(nodeType) {
    const nodeKey = `node:${nodeType}`;

    if (!this.signature[nodeKey]) {
      return false;
    }

    // Remove all sections for this node
    const keysToRemove = Object.keys(this.signature).filter(
      key => key.startsWith(`node:${nodeType}`)
    );

    for (const key of keysToRemove) {
      delete this.signature[key];
    }

    // Update metadata
    if (this.signature.metadata) {
      this.signature.metadata.authenticated_nodes =
        Math.max(0, (this.signature.metadata.authenticated_nodes || 1) - 1);
    }

    await this.save();
    return true;
  }

  /**
   * Update default parameters for a node
   */
  async updateNodeDefaults(nodeType, defaults) {
    if (!this.isAuthenticated(nodeType)) {
      return false;
    }

    const defaultsKey = `node:${nodeType}.defaults`;
    this.signature[defaultsKey] = defaults;

    await this.save();
    return true;
  }

  /**
   * Get list of all authenticated node types
   */
  getAuthenticatedNodes() {
    const nodes = [];
    for (const key of Object.keys(this.signature)) {
      if (key.startsWith('node:') && !key.includes('.')) {
        const nodeType = key.split(':')[1];
        if (this.isAuthenticated(nodeType)) {
          nodes.push(nodeType);
        }
      }
    }
    return nodes;
  }

  /**
   * Validate signature file structure
   */
  validate() {
    const errors = [];
    const warnings = [];

    // Check for required sections
    if (!this.signature.signature) {
      errors.push('Missing [signature] section');
    } else {
      if (!this.signature.signature.version) {
        errors.push('Missing signature.version');
      }
    }

    // Check node sections
    for (const key of Object.keys(this.signature)) {
      if (key.startsWith('node:') && !key.includes('.')) {
        const nodeType = key.split(':')[1];

        // Check for required fields
        if (this.signature[key].authenticated === undefined) {
          errors.push(`Node '${nodeType}' missing 'authenticated' field`);
        }

        // Check auth section if authenticated
        if (this.signature[key].authenticated) {
          const authKey = `node:${nodeType}.auth`;
          if (!this.signature[authKey]) {
            errors.push(`Node '${nodeType}' is authenticated but missing .auth section`);
          } else {
            // Check env references
            const auth = this.signature[authKey];
            for (const [field, value] of Object.entries(auth)) {
              if (typeof value === 'string' && value.includes('{{.env.')) {
                const envVar = this._extractEnvVar(value);
                if (envVar && !process.env[envVar]) {
                  warnings.push(`Environment variable '${envVar}' not set for ${nodeType}.${field}`);
                }
              }
            }
          }
        }
      }
    }

    return {
      valid: errors.length === 0,
      errors,
      warnings,
      authenticated_nodes: this.getAuthenticatedNodes().length
    };
  }

  /**
   * Get complete signature information
   */
  getSignatureInfo(nodeType = null) {
    if (nodeType) {
      // Return info for specific node
      if (!this.isAuthenticated(nodeType)) {
        return {
          authenticated: false,
          error: `Node '${nodeType}' is not authenticated`
        };
      }

      const operations = this.getOperations(nodeType);
      return {
        type: nodeType,
        authenticated: true,
        operations: Object.keys(operations),
        operation_count: Object.keys(operations).length,
        defaults: this.getNodeDefaults(nodeType),
        metadata: this.getNodeMetadata(nodeType)
      };
    }

    // Return info for all nodes
    const authenticated = this.getAuthenticatedNodes();

    return {
      version: this.signature.signature?.version || 'unknown',
      user_id: this.signature.signature?.user_id || 'unknown',
      authenticated_nodes: authenticated.map(nodeType => ({
        type: nodeType,
        operations: Object.keys(this.getOperations(nodeType)),
        operation_count: Object.keys(this.getOperations(nodeType)).length,
        defaults: this.getNodeDefaults(nodeType),
        metadata: this.getNodeMetadata(nodeType)
      })),
      total_authenticated: authenticated.length,
      updated_at: this.signature.signature?.updated_at || 'unknown'
    };
  }

  /**
   * Resolve {{.env.VARIABLE}} references recursively
   */
  _resolveEnvVars(obj) {
    if (typeof obj === 'object' && obj !== null) {
      if (Array.isArray(obj)) {
        return obj.map(item => this._resolveEnvVars(item));
      } else {
        const resolved = {};
        for (const [key, value] of Object.entries(obj)) {
          resolved[key] = this._resolveEnvVars(value);
        }
        return resolved;
      }
    } else if (typeof obj === 'string' && obj.includes('{{.env.')) {
      const envVar = this._extractEnvVar(obj);
      if (envVar) {
        return process.env[envVar] || obj;
      }
    }
    return obj;
  }

  /**
   * Extract environment variable name from {{.env.VAR}} syntax
   */
  _extractEnvVar(value) {
    const match = value.match(/\{\{\.env\.(\w+)\}\}/);
    return match ? match[1] : null;
  }

  /**
   * Convert auth values to {{.env.VARIABLE}} references
   */
  _toEnvReferences(nodeType, auth) {
    const result = {};
    for (const [key, value] of Object.entries(auth)) {
      const envVarName = `${nodeType.toUpperCase()}_${key.toUpperCase()}`;
      result[key] = `{{.env.${envVarName}}}`;
    }
    return result;
  }

  /**
   * Test authentication with external API
   */
  async validateAuth(nodeType, authData) {
    // Node-specific validation
    if (nodeType === 'github') {
      const response = await fetch('https://api.github.com/user', {
        headers: {
          'Authorization': `Bearer ${authData.access_token}`,
          'User-Agent': 'Flow-Architect-MCP'
        }
      });
      if (!response.ok) {
        throw new Error('Invalid GitHub token');
      }
      return await response.json();
    }

    if (nodeType === 'openai') {
      const response = await fetch('https://api.openai.com/v1/models', {
        headers: {
          'Authorization': `Bearer ${authData.api_key}`
        }
      });
      if (!response.ok) {
        throw new Error('Invalid OpenAI API key');
      }
      return await response.json();
    }

    // Add more node types as needed

    // For unknown node types, just return true
    return { validated: true, message: 'No validation available for this node type' };
  }
}
